Zend/zend_API.c

```diff
@@ -1627,13 +1627,31 @@ ZEND_API void object_properties_load(zend_object *object, HashTable *properties)
 					zend_hash_update(object->properties, key, &tmp);
 				}
 			} else {
+				if (UNEXPECTED(object->ce->ce_flags & ZEND_ACC_NO_DYNAMIC_PROPERTIES)) {
+					zend_throw_error(NULL, "Cannot create dynamic property %s::$%s",
+						ZSTR_VAL(object->ce->name), zend_get_unmangled_property_name(key));
+					return;
+				} else if (!(object->ce->ce_flags & ZEND_ACC_ALLOW_DYNAMIC_PROPERTIES)) {
+					zend_error(E_DEPRECATED, "Creation of dynamic property %s::$%s is deprecated",
+						ZSTR_VAL(object->ce->name), zend_get_unmangled_property_name(key));
+				}
+
 				if (!object->properties) {
 					rebuild_object_properties(object);
 				}
 				prop = zend_hash_update(object->properties, key, prop);
 				zval_add_ref(prop);
 			}
 		} else {
+			if (UNEXPECTED(object->ce->ce_flags & ZEND_ACC_NO_DYNAMIC_PROPERTIES)) {
+				zend_throw_error(NULL, "Cannot create dynamic property %s::$%s",
+					ZSTR_VAL(object->ce->name), zend_get_unmangled_property_name(key));
```

Comment:
This is wrong - this is in the else branch where `if (key)` is false, i.e. `key` is null.

Instead, change this to 

```suggestion
				zend_throw_error(NULL, "Cannot create dynamic property %s::$" ZEND_LONG_FMT,
					ZSTR_VAL(object->ce->name), h);
```

Is there a way to test numbers in the dynamic property hash table with a readonly class and one of the helpers (e.g. spl) using `object_properties_load`

Reply:
```
php > $g = new GMP();
php > $g->{1} = 123;

Deprecated: Creation of dynamic property GMP::$1 is deprecated in php shell code on line 1
php > echo serialize($g);
O:3:"GMP":2:{i:0;s:1:"0";i:1;a:1:{i:1;i:123;}}
php > var_export(unserialize(serialize($g)));
\GMP::__set_state(array(
   1 => 123,
))
```

e.g. this should test the deprecation case when `key` is null, since gmp uses object_properties_load?